framework/auth/core.py

```diff
@@ -541,14 +596,47 @@ def _set_email_token_expiration(self, token, expiration=None):
         self.email_verifications[token]['expiration'] = expiration
         return expiration
 
-    def add_email_verification(self, email, expiration=None):
+    def add_unconfirmed_email(self, email, expiration=None):
         """Add an email verification token for a given email."""
+        # TODO: If the unconfirmed email is already present, refresh the token
+
+        # TODO: This is technically not compliant with RFC 822, which requires
+        #       that case be preserved in the "local-part" of an address. From
+        #       a practical standpoint, the vast majority of email servers do
+        #       not preserve case.
+        #       ref: https://tools.ietf.org/html/rfc822#section-6
+        email = email.lower().strip()
+
+        validate_email(email)
+
         token = generate_confirm_token()
 
-        self.email_verifications[token] = {'email': email.lower()}
+        self.email_verifications[token] = {'email': email}
         self._set_email_token_expiration(token, expiration=expiration)
         return token
 
+    def remove_unconfirmed_email(self, email):
+        """Remove an unconfirmed email addresses and their tokens."""
+        for token, value in self.email_verifications.iteritems():
+            if value.get('email') == email:
+                del self.email_verifications[token]
+                return True
+
+        return False
+
+    def remove_email(self, email):
+        """Remove a confirmed email"""
+        if email in self.emails:
+            self.emails.remove(email)
+            mails.send_mail(self.username,
```

Comment:
Might make sense to send these emails in a signal subscriber. You should be able to make a method a subscriber:

```python
    def remove_email(self):
        self.email.remove(email)
        user_email_removed.send(self, email)

    @user_email_removed.connect
    def send_email_removal_confirmations(self, email):
        if email in self.emails:
            #...
```